### PR TITLE
feat(proxy): support sending requests through HTTPS_PROXY

### PR DESCRIPTION
**Pull request by mithlesh135**
## Describe your changes
For external providers generally access is via a proxy, currently that is missing, this PR add the capability to set http_proxy or https_proxy in env and app will take care of proxying the request via this tunnel

## Issue ticket number and link
#2233
## Checklist before requesting a review (skip if just adding/editing APIs & templates)
- [ ] I added tests, otherwise the reason is: 
- [ ] I added observability, otherwise the reason is:
- [ ] I added analytics, otherwise the reason is:

### CODE REVIEW

**Review by khaliqgant — APPROVED** [review]

**Review by khaliqgant — APPROVED** [review]

**Comment by bodinsamuel** [comment]
Is HTTP (no "s") a requirement for you? 
I don't think we should support that because it's a liability for credentials and potential security breach.

**Review by TBonnin — CHANGES_REQUESTED** [review]
I think https://github.com/NangoHQ/nango/blob/6581d8a3114e7df3ed383907ac9db542eeae5f41/packages/shared/lib/services/proxy.service.ts should also be modified to take into account the proxy env var

**Comment by mithlesh135** [comment]
> Is HTTP (no "s") a requirement for you? I don't think we should support that because it's a liability for credentials and potential security breach.

So the proxy will be a forward proxy which will sit on our side, so application can talk to this proxy on http, this proxy will be talking on https with external client

**Comment by khaliqgant** [comment]
> I think https://github.com/NangoHQ/nango/blob/6581d8a3114e7df3ed383907ac9db542eeae5f41/packages/shared/lib/services/proxy.service.ts should also be modified to take into account the proxy env var

If they're running syncs and actions, right? Which I don't think they would be since this is self hosted.

**Comment by mithlesh135** [comment]
> I think https://github.com/NangoHQ/nango/blob/6581d8a3114e7df3ed383907ac9db542eeae5f41/packages/shared/lib/services/proxy.service.ts should also be modified to take into account the proxy env var

Done

**Comment by mithlesh135** [comment]
> > I think https://github.com/NangoHQ/nango/blob/6581d8a3114e7df3ed383907ac9db542eeae5f41/packages/shared/lib/services/proxy.service.ts should also be modified to take into account the proxy env var
> 
> If they're running syncs and actions, right? Which I don't think they would be since this is self hosted.

I have added it to this file as well

**Comment by TBonnin** [comment]
> > > I think https://github.com/NangoHQ/nango/blob/6581d8a3114e7df3ed383907ac9db542eeae5f41/packages/shared/lib/services/proxy.service.ts should also be modified to take into account the proxy env var
> > 
> > 
> > If they're running syncs and actions, right? Which I don't think they would be since this is self hosted.
> 
> I have added it to this file as well

Thank you. Even though you might not use it if you are not running any integration scripts, it is good to be cover all use cases so people running enterprise releases can also use those env vars

**Review by TBonnin — APPROVED** [review]

**Comment by bodinsamuel** [comment]
> So the proxy will be a forward proxy which will sit on our side, so application can talk to this proxy on http, this proxy will be talking on https with external client

Not sure I understand your answer. Also I'm not entirely getting the scope, should all calls to the external API goes through your Proxy or just the proxy one? If the former then it's missing some spots.


Here is a proposal for this PR https://github.com/NangoHQ/nango/compare/proposal-http-agent?expand=1

Differences:
- Renamed to utils/axios.ts
- No support for HTTP proxy (unless I'm missing something, we should never reach anything without encryption)
- Support for regular agent (
- Missing connection file (@khaliqgant can you double check)

**Comment by khaliqgant** [comment]
> * Missing connection file (@khaliqgant can you double check)

Yeah, good call on the connection.service file. For that matter if it is all outgoing requests then the `webhook.service.ts` file should be updated as well.

**Comment by mithlesh135** [comment]
> > So the proxy will be a forward proxy which will sit on our side, so application can talk to this proxy on http, this proxy will be talking on https with external client
> 
> Not sure I understand your answer. Also I'm not entirely getting the scope, should all calls to the external API goes through your Proxy or just the proxy one? If the former then it's missing some spots.
> 
> Here is a proposal for this PR https://github.com/NangoHQ/nango/compare/proposal-http-agent?expand=1
> 
> Differences:
> 
> * Renamed to utils/axios.ts
> * No support for HTTP proxy (unless I'm missing something, we should never reach anything without encryption)
> * Support for regular agent (
> * Missing connection file (@khaliqgant can you double check)

Taken care for most of proposals, let me know if I should rename as well

**Comment by mithlesh135** [comment]
> > * Missing connection file (@khaliqgant can you double check)
> 
> Yeah, good call on the connection.service file. For that matter if it is all outgoing requests then the `webhook.service.ts` file should be updated as well.

added both the files

**Comment by mithlesh135** [comment]
> https://github.com/NangoHQ/nango/compare/proposal-http-agent?expand=1

@bodinsamuel   Support for http-proxy will be needed, most of these kind of proxies connect to own apps via http only as app and this proxy will sit in same network so its fine to communicate over http, now this proxy will be making https calls to external provider


Screenshot 2024-06-03 at 6 58 44 PM

File: packages/server/lib/controllers/proxy.controller.ts [hunk 1]
```diff
@@ -7,7 +7,7 @@ import type { UrlWithParsedQuery } from 'url';
 import url from 'url';
 import querystring from 'querystring';
 import type { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios';
-import axios from 'axios';
+import { axiosInstance as axios } from '@nangohq/shared/lib/axiosInstance/index.js';
```
Comment (bodinsamuel) [inline packages/server/lib/controllers/proxy.controller.ts]: import is wrong, it should be export from shared index.ts

File: packages/shared/lib/axiosInstance/index.ts
Comment (bodinsamuel) [inline packages/shared/lib/axiosInstance/index.ts:1]: would still put it in utils, to avoid having one folder per util

**Review by bodinsamuel — COMMENTED** [review]
I now see the point with the graph thanks 👌🏻 🙏🏻 

Still believe it's a footgun but we'll see. Added a few comments, otherwise looks good. I'll apply my change regarding the agent in an other PR 👍🏻

File: packages/shared/lib/axiosInstance/index.ts
Comment (mithlesh135) [reply packages/shared/lib/axiosInstance/index.ts:1]: ok, let me move it to utils then

Comment (mithlesh135) [reply packages/shared/lib/axiosInstance/index.ts:1]: not getting one folder per util ?

File: packages/server/lib/controllers/proxy.controller.ts (hunk 1 above)
Comment (mithlesh135) [reply packages/server/lib/controllers/proxy.controller.ts]: fixed

File: packages/shared/lib/axiosInstance/index.ts
Comment (mithlesh135) [reply packages/shared/lib/axiosInstance/index.ts:1]: @bodinsamuel Done (moved to utils)

File: packages/shared/lib/clients/oauth2.client.ts [hunk 2]
```diff
@@ -93,10 +111,14 @@ export async function getFreshOAuth2Credentials(
         }
 
         if (Boom.isBoom(e)) {
-            const payload = { external_message: e.message, external_request_details: e.output, dataMessage: errorPayload };
+            const payload = {
+                external_message: e.message,
+                external_request_details: e.output,
+                dataMessage: errorPayload instanceof Buffer ? errorPayload.toString() : errorPayload
```
Comment (bodinsamuel) [inline packages/shared/lib/clients/oauth2.client.ts:117]: This would output raw buffer

File: packages/shared/lib/clients/oauth2.client.ts [hunk 3]
```diff
@@ -93,10 +111,14 @@ export async function getFreshOAuth2Credentials(
...
+            };
             nangoErr = new NangoError(`refresh_token_external_error`, payload);
         } else {
-            nangoErr = new NangoError(`refresh_token_external_error`, { message: errorPayload });
+            nangoErr = new NangoError(`refresh_token_external_error`, { message: e.message });
```
Comment (bodinsamuel) [inline packages/shared/lib/clients/oauth2.client.ts:121]: This would output raw buffer

File: packages/shared/lib/services/connection.service.ts [hunk 4]
```diff
@@ -702,6 +702,7 @@ class ConnectionService {
                     });
                 }
 
+                // TODO: this leak credentials to the logs
                 const errorWithPayload = new NangoError(error.type, connection);
 
                 return Err(errorWithPayload);
```
Comment (bodinsamuel) [inline packages/shared/lib/services/connection.service.ts:708]: This return is not nice, it's leaking credentials and also force the API to return a 400 which is not true

**Review by bodinsamuel — COMMENTED** [review]
@TBonnin @khaliqgant 
Can you test locally please? I can't manage to refresh on my side I keep getting `incomplete_raw_credentials` for some reason.

**Review by khaliqgant — APPROVED** [review]
Tested this and works as expected for normal operations with Nango (proxy, auth call etc)

**Comment by khaliqgant** [comment]
> @TBonnin @khaliqgant Can you test locally please? I can't manage to refresh on my side I keep getting `incomplete_raw_credentials` for some reason.

Noting here that if you were testing with Github and forcing a refresh in the code it is going to give the `incomplete_raw_credentials` response because Github doesn't issue a refresh token

**Comment by bodinsamuel** [comment]
> Noting here that if you were testing with Github and forcing a refresh in the code it is going to give the incomplete_raw_credentials response because Github doesn't issue a refresh token

Alright, thanks a lot for re-testing.

**Pull request merged**